components/tidb_query_datatype/src/codec/data_type/chunked_vec_enum.rs

```diff
@@ -3,41 +3,50 @@
 use super::bit_vec::BitVec;
 use super::{ChunkRef, ChunkedVec, UnsafeRefInto};
 use super::{Enum, EnumRef};
+use crate::codec::data_type::{ChunkedVecBytes, ChunkedVecSized, Int};
 use crate::impl_chunked_vec_common;
-use std::sync::Arc;
-use tikv_util::buffer_vec::BufferVec;
 
-/// `ChunkedVecEnum` stores enum in a compact way.
+/// `ChunkedVecEnum` is a vector storing `Option<Enum>`.
 ///
 /// Inside `ChunkedVecEnum`:
-/// - `data` stores the real enum data.
-/// - `bitmap` indicates if an element at given index is null.
-/// - `value` is an 1-based index enum data offset, 0 means this enum is ''
+/// - `values` stores 1-based index enum data offsets, 0 means this enum is ''
+/// - `names` stores the names of enum data
 ///
 /// # Notes
 ///
-/// Make sure operating `bitmap` and `value` together to prevent different
-/// stored representation issue discussed at
-/// <https://github.com/tikv/tikv/pull/8948#discussion_r516463693>
+/// `values` and `names` both maintain a duplicated bitmap.
+/// We are not able to store the data in a more compact form
+/// because we have to borrow the reference of `ChunkedVecSize<Int>`
+/// and `ChunkedVecBytes`. The borrowed reference should have the same
+/// lifetime with `ChunkedVecEnum` which cannot be achieved if we don't
+/// store the owned data in struct fields.
 ///
-/// TODO: add way to set enum column data
+/// TODO: Change to a compact format when TiDB has its Enum/Set EvalType
 #[derive(Debug, Clone)]
 pub struct ChunkedVecEnum {
-    data: Arc<BufferVec>,
-    bitmap: BitVec,
     // MySQL Enum is 1-based index, value == 0 means this enum is ''
-    value: Vec<usize>,
+    values: ChunkedVecSized<Int>,
+    names: ChunkedVecBytes,
 }
```

Comment:
This change may cause increase of memory usage. Duplicated **bitmap** in `ChunkVecSized<Int>` and `ChunkedVecByte`.

Also origin implementation use pointer to store name, here we store real enum definition.

Maybe we can find another way to fix problem in `borrow_vector_ref()`

Reply:
> Also origin implementation use pointer to store name, here we store real enum definition.

This has been well-discussed before. Sticking to current design means that we had to store each value in both `Int` and `Bytes` format sequentially, instead of using something like pointer.

> This change may cause increase of memory usage. Duplicated bitmap in `ChunkVecSized<Int>` and `ChunkedVecByte`.

Agree with this. To make things simple, I suggest using a `Vec<Int>` and a `ChunkedVecByte` here.

Reply:
The issue lies in the design. In current PR, enum could be implicitly cast to bytes and int. Therefore, we could not reduce this memory footprint.